miniapps/adjoint/cvsRoberts_ASAi_dns.cpp

```diff
@@ -0,0 +1,385 @@
+      w.Print();
+
+      cvodes->GetForwardSolution(t, u);
+      cout << "u:" << endl;
+      u.Print();
+
+      // Results at T0
+      dt_real = max(dt, t - 0.);
+      cvodes->StepB(w, t, dt_real);
+      cout << "t: " << t << endl;
+      cout << "w:" << endl;
+
+      cvodes->GetForwardSolution(t, u);
+      w.Print();
+      cout << "u:" << endl;
+      u.Print();
+
+      // Evaluate Sensitivity
+      cvodes->EvalQuadIntegrationB(t, dG_dp);
+      cout << "dG/dp:" << endl;
+      dG_dp.Print();
+
+   }
+
+   // 8. Free the used memory.
+   delete cvodes;
+
+   return 0;
+}
+
+// cvsRoberts_ASAi_dns rate equation
+void RobertsTDAOperator::Mult(const Vector &x, Vector &y) const
+{
+   y[0] = -p_[0]*x[0] + p_[1]*x[1]*x[2];
+   y[2] = p_[2]*x[1]*x[1];
+   y[1] = -y[0] - y[2];
+}
+
+// cvsRoberts_ASAi_dns quadrature rate equation
+void RobertsTDAOperator::QuadratureIntegration(const Vector &y,
+					       Vector &qdot) const
+{
+   qdot[0] = y[2];
+}
+
+// cvsRoberts_ASAi_dns adjoint rate equation
+void RobertsTDAOperator::AdjointRateMult(const Vector &y, Vector & yB,
+                                      Vector &yBdot) const
+{
+   double l21 = (yB[1]-yB[0]);
+   double l32 = (yB[2]-yB[1]);
+   double p1 = p_[0];
+   double p2 = p_[1];
+   double p3 = p_[2];
+   yBdot[0] = -p1 * l21;
+   yBdot[1] = p2 * y[2] * l21 - 2. * p3 * y[1] * l32;
+   yBdot[2] = p2 * y[1] * l21 - 1.0;
+}
+
+// cvsRoberts_ASAi_dns quadrature sensitivity rate equation
+void RobertsTDAOperator::QuadratureSensitivityMult(const Vector &y,
+						   const Vector &yB,
+						   Vector &qBdot) const
+{
+   double l21 = (yB[1]-yB[0]);
+   double l32 = (yB[2]-yB[1]);
+   double y23 = y[1] * y[2];
+
+   qBdot[0] = y[0] * l21;
+   qBdot[1] = -y23 * l21;
+   qBdot[2] = y[1]*y[1]*l32;
+}
+
+// cvsRoberts_ASAi_dns implicit solve setup for adjoint
+int RobertsTDAOperator::SUNImplicitSetupB(const double t, const Vector &y,
+					  const Vector &yB,
+					  const Vector &fyB, int jokB,
+					  int *jcurB, double gammaB)
+{
+
+   // M = I- gamma J
+   // J = dfB/dyB
+   // Let's create a SparseMatrix and fill in the entries since this example
+   // doesn't contain finite elements
+
+   delete adjointMatrix;
+   adjointMatrix = new SparseMatrix(y.Size(), yB.Size());
+   for (int j = 0; j < y.Size(); j++)
+   {
+      Vector JacBj(yB.Size());
+      Vector yBone(yB.Size());
+      yBone = 0.;
+      yBone[j] = 1.;
+      AdjointRateMult(y, yBone, JacBj);
+      JacBj[2] += 1.;
+      for (int i = 0; i < y.Size(); i++)
+      {
+         adjointMatrix->Set(i,j, (i == j ? 1.0 : 0.) - gammaB * JacBj[i]);
+      }
+   }
+
+   *jcurB = 1;
+   adjointMatrix->Finalize();
+   adjointSolver.SetOperator(*adjointMatrix);
+
+   return 0;
+}
+
+// cvsRoberts_ASAi_dns implicit solve for adjoint 
+int RobertsTDAOperator::SUNImplicitSolveB(Vector &x, const Vector &b,
+					  double tol)
+{
+  // The tolerance is ignored in this example as we're trying to replicate
+  // CVODES cvsRoberts_ASAi_dns
```

Comment:
I don't see how tolerance here is ignored. Tolerance is very important here since you are looking for a direct solve to replicate the direct numerical solve in the original CVODES example. On a sidenote, since the problem is using SparseMatrix you can use the LU factorization to achieve a direct solve instead of using a Krylov accelerator like GMRES.

Reply:
The "tol" is unused is the point I was trying to get across.

Reply:
Oh you were referring to the tolerance in the function.

Reply:
Yes, I'll be a little more specific since I didn't want MFEM users to wonder why changing some of the sundials interface tolerances won't affect the tolerance of the solution in the example.